### PR TITLE
Update metering to version 1.0

### LINKED ISSUE #9856: Update Metering
Metering was updated to a new major version which changes the deployment and introduce a dependency to prom

Acceptance criteria:
Update Metering Deployment to the new Version

### PR DESCRIPTION
**Pull request by mate4st**
**What this PR does / why we need it**:

Removes  the legacy components of metering and deploys version 1.0 of metering.
This also introduce multiple types of reports, cluster and namespaced based. 
By default both types are generated, but it can be configured per report configuration. 

**Which issue(s) this PR fixes**:

Fixes #9856 

**What type of PR is this?**
/kind feature

**Special notes for your reviewer**:

Since all data of legacy metering installation are synced in 5minutes intervals, we can safely delete the existing pvc.

**Does this PR introduce a user-facing change? Then add your Release Note here**:

```release-note
[Breaking] Update Metering to version 1.0. This changes the report csv format and the method of data collection. All previous generated reports will be still accessible via the dashboard. The new metering has a slightly different format and uses a different data source. This means that data collection will start from the beginning at the time of upgrading. Going back in time is not possible due to the change of the data source.
```

**Documentation**:

```documentation
https://github.com/kubermatic/docs/pull/1160
```

### CODE REVIEW

**Comment by xrstf** [comment]
> The new metering installation should be in its own namespace

.... because? :) Please describe your reasonings for future reference.

**Comment by WeirdMachine** [comment]
> > The new metering installation should be in its own namespace
> 
> .... because? :) Please describe your reasonings for future reference.



> > The new metering installation should be in its own namespace
> 
> .... because? :) Please describe your reasonings for future reference.

I think it make sense to have it in its own namespace, because its a more or less independent feature that could be seperated, similar to mla. However during testing I found out that this breaks reconcile where we use the seed owner reference because its not allowed to be cross namespaces. Therefore I reverted the namespace seperation and moved metering back to kubermatic namesapce.

File: pkg/ee/metering/prometheus/clusterole.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+package prometheus
+
+import (
+	"k8c.io/kubermatic/v2/pkg/controller/operator/common"
+	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+)
+
+// clusterRoleBindingCreator create a cluster role binding for the metering tool.
```
Comment (xrstf) [inline pkg/ee/metering/prometheus/clusterole.go]: Wrong comment

File: pkg/ee/metering/prometheus/clusterrolebinding.go [hunk 2]
```diff
@@ -22,28 +22,33 @@
...
+package prometheus
 
 import (
+	"k8c.io/kubermatic/v2/pkg/controller/operator/common"
 	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
 
 	rbacv1 "k8s.io/api/rbac/v1"
 )
 
 // clusterRoleBindingCreator create a cluster role binding for the metering tool.
```
Comment (xrstf) [inline pkg/ee/metering/prometheus/clusterrolebinding.go]: Wrong comment

File: pkg/ee/metering/prometheus/configmap.go [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+package prometheus
+
+import (
+	"k8c.io/kubermatic/v2/pkg/controller/operator/common"
+	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+// prometheusStatefulSe creates the service account used by prometheus.
```
Comment (xrstf) [inline pkg/ee/metering/prometheus/configmap.go]: Wrong comment.

File: pkg/ee/metering/prometheus/reconcile.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	Namespace = resources.KubermaticNamespace
+)
+
+// ReconcilePrometheus reconciles the prometheus instance used as a datasource for metering.
+func ReconcilePrometheus(ctx context.Context, client ctrlruntimeclient.Client, scheme *runtime.Scheme, getRegistry registry.WithOverwriteFunc, seed *kubermaticv1.Seed) error {
+	seedOwner := common.OwnershipModifierFactory(seed, scheme)
+
+	if err := persistentVolumeClaimCreator(ctx, client, seed); err != nil {
+		return fmt.Errorf("failed to reconcile PVC: %w", err)
+	}
```
Comment (xrstf) [inline pkg/ee/metering/prometheus/reconcile.go]: Why do we manually create a single PVC that is then used by potentially multiple Prometheus pods from the STS? Shouldn't the STS simply have a PVC template attached?

File: pkg/ee/metering/prometheus/svc.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+import (
+	"k8c.io/kubermatic/v2/pkg/controller/operator/common"
+	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// prometheusStatefulSe creates the service account used by prometheus.
```
Comment (xrstf) [inline pkg/ee/metering/prometheus/svc.go]: Wrong comment.

File: pkg/ee/metering/reconcile.go [hunk 6]
```diff
@@ -179,50 +176,115 @@ func cleanupOrphanedReportingCronJobs(ctx context.Context, client ctrlruntimecli
...
-		return fmt.Errorf("failed to cleanup metering Deployment: %w", err)
+// fetchExistingReportingCronJobs returns a list of all existing reporting cronjobs.
+func fetchExistingReportingCronJobs(ctx context.Context, client ctrlruntimeclient.Client) (*batchv1.CronJobList, error) {
+	existingReportingCronJobs := &batchv1.CronJobList{}
+	listOpts := []ctrlruntimeclient.ListOption{
+		ctrlruntimeclient.InNamespace(meteringNamespace),
+		ctrlruntimeclient.ListOption(ctrlruntimeclient.MatchingLabels{common.ComponentLabel: meteringName}),
+	}
+	if err := client.List(ctx, existingReportingCronJobs, listOpts...); err != nil {
+		if !apierrors.IsNotFound(err) {
```
Comment (xrstf) [inline pkg/ee/metering/reconcile.go:186]: Can a 404 really happen when listing CronJobs?

File: pkg/ee/metering/reconcile.go [hunk 7]
```diff
@@ -179,50 +176,115 @@ func cleanupOrphanedReportingCronJobs(ctx context.Context, client ctrlruntimecli
...
+			return nil, fmt.Errorf("failed to list reporting cronjobs: %w", err)
+		}
 	}
+	return existingReportingCronJobs, nil
+}
+
+// undeploy removes all metering components expect the namespace and pvc.
```
Comment (xrstf) [inline pkg/ee/metering/reconcile.go]: s/expect/except/

File: pkg/ee/metering/reconcile.go [hunk 8]
```diff
@@ -179,50 +176,115 @@ func cleanupOrphanedReportingCronJobs(ctx context.Context, client ctrlruntimecli
...
+	}
+	if err := cleanupResource(ctx, client, key, &rbacv1.ClusterRoleBinding{}); err != nil {
+		return fmt.Errorf("failed to cleanup metering pull secret: %w", err)
+	}
+	if err := cleanupResource(ctx, client, key, &rbacv1.ClusterRole{}); err != nil {
+		return fmt.Errorf("failed to cleanup metering pull secret: %w", err)
+	}
+	if err := cleanupResource(ctx, client, key, &corev1.ServiceAccount{}); err != nil {
+		return fmt.Errorf("failed to cleanup metering pull secret: %w", err)
+	}
```
Comment (xrstf) [inline pkg/ee/metering/reconcile.go]: All of those error messages are wrong.

File: pkg/ee/metering/reconcile.go (hunk 6 above)
Comment (mate4st) [reply pkg/ee/metering/reconcile.go:186]: I assume this happens when parallel reconcile is enabled or someone manually deleting it at the same time. However we can always ignore this error.

File: pkg/ee/metering/prometheus/reconcile.go (hunk 4 above)
Comment (mate4st) [reply pkg/ee/metering/prometheus/reconcile.go]: Did so. Is there anything i need to take in consideration when using those?

File: pkg/apis/kubermatic/v1/datacenter.go [hunk 9]
```diff
@@ -773,6 +773,9 @@ type SeedMLASettings struct {
 // MeteringConfiguration contains all the configuration for the metering tool.
 type MeteringConfiguration struct {
 	Enabled bool `json:"enabled"`
+
+	// +kubebuilder:default=kubermatic-fast
```
Comment (xrstf) [inline pkg/apis/kubermatic/v1/datacenter.go]: Let's not hardcode this value into our CRDs. If anything we should do the defaulting in our own code and make this maybe dependent on the KubermaticConfiguration. "kubermatic-fast" was also originally only meant for SSD volumes for etcds, not for "everthing that KKP stores", so `kubermatic-fast` might even be the wrong default here.

I would just use the default StorageClass from the seed cluster, unless the admin configures a special SC in the Seed datacenter.

File: pkg/ee/metering/prometheus/reconcile.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+	"k8c.io/kubermatic/v2/pkg/resources/registry"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	Name      = "metering-prometheus"
+	Namespace = resources.KubermaticNamespace
```
Comment (xrstf) [inline pkg/ee/metering/prometheus/reconcile.go]: Please use the `seed.Namespace` instead of hardcoding it into the controller.

**Review by xrstf — APPROVED** [review]
/approve

**Comment by pkprzekwas** [comment]
In order to add support for types in the dashboard we need to extend report configuration handler, but it can be done in a separate PR.

**Review by pkprzekwas — APPROVED** [review]

File: pkg/ee/metering/cronjob.go [hunk 11]
```diff
@@ -38,136 +40,33 @@ import (
...
+				job.Labels = make(map[string]string)
 			}
-			labels[LabelKey] = reportName
-			job.SetLabels(labels)
+			job.Labels[common.NameLabel] = reportName
+			job.Labels[common.ComponentLabel] = meteringName
 
 			job.Spec.Schedule = mrc.Schedule
-			job.Spec.JobTemplate.Spec.Parallelism = pointer.Int32Ptr(1)
-			job.Spec.JobTemplate.Spec.Template.Spec.ServiceAccountName = meteringToolName
```
Comment (wozniakjan) [inline pkg/ee/metering/cronjob.go:53]: I think this causes an issue for upgraded clusters, the fact `ServiceAccountName` is not explicitly removed causes old `CronJobs` to contain the old value and that `ServiceAccount` is purged as a legacy resource in the mechanism introduced in this PR.
https://github.com/kubermatic/kubermatic/blob/8fd688856e73375558ecb9969e34388158d0aff6/pkg/ee/metering/reconcile.go#L251-L253

**Pull request merged**